### ISSUE TITLE
Changing func syntax

### ISSUE THREAD

**Issue by max-sixty** [PRQL/prql#2504]
### What's up?

This has been brewing for some time and it might be worth doing it now. Currently the syntax is a bit "too functional', with no separation between the function name and its parameters, making it less consistent with the rest of the language. Mostly this is an artifact of history — we had functions before we had `let`.

Currently:

```elm
func f temp -> (temp - 32) / 1.8
```

A couple options — relatively small differences between them. All add an `=`.
- Lambda syntax with `func`: 
  ```elm
  func f = temp -> (temp - 32) / 1.8
  ```
  This makes `let` orthogonal — we can declare anything using it, including functions.
- Lambda syntax with `let`:
  ```elm
  let f = temp -> (temp - 32) / 1.8
  ```
  This is more consistent, but harder to quickly see where is a function vs a relation declaration. Though it would be possible to type the `let` declaration:
  ```diff
  -let f = temp -> (temp - 32) / 1.8
  +let f<func> = temp -> (temp - 32) / 1.8
  ```
- Allow both forms. Because if we only go with the former, it makes `let` not fully orthogonal, which is maybe bad? But then we have two ways of doing something...

Any other options or thoughts?

**Comment by vanillajonathan** [comment]
> There should be one-- and preferably only one --obvious way to do it.

From the [Zen of Python](https://peps.python.org/pep-0020/).

The C# language uses `=>` instead of `->` for expression-bodied functions and lambdas. Example:
```elm
func f temp => (temp - 32) / 1.8
```

This also needs to work with:
- functions that accept multiple parameters
- functions that span multiple lines
- parameters with type hints
- parameters with default values
- functions with optional parameters
- return type with type hints

A clearer syntax might be:
```elm
func fahrenheit(temp: int) => (temp - 32) / 1.8
```

Keep in mind that angel brackets are easily confused with generics.

**Comment by max-sixty** [comment]
I was thinking we'd make a fairly small change just to align the syntax. I'm ofc open to broader changes in the future, but I still think we should keep functions very simple for the moment — I don't see a clear need to add complexity, or non-orthogonal features such as multiple lines (https://github.com/PRQL/prql/issues/819 can do this!)

I also don't think something python-like, e.g. `func fahrenheit(temp: int)` would fit with the rest of the language — we don't use parentheses for calling, we use colons to pass args, etc.

I agree angle brackets look a bit like generics, open to changing those (though unclear what we'd chose — we already use colons). We can also change those separately.

**Comment by vanillajonathan** [comment]
Colon would be ideal for type hinting since colon is widely used in many other languages for type hinting.

**Comment by aljazerzen** [comment]
Unfortunately, colon and `=>` are already used for other things and we did already have quite a lot of discussion around this.

@max-sixty I'm all for the proposed change - the lambda notation is needed for other things too, so that is a must.

I think that the `func` notation can also be changed as you propose, to be more consistent. If anyone has strong opinions regarding "single one to do it" here, we can remove it in favor of lambda notation. But I think this is just a syntactic sugar for convenience and because one can easily convert between the two, this is not the violation of the rule.

We can also extend `func` notation to support parameters on multiple lines (as @vanillajonathan noted), which we cannot do for lambda functions as that would break new-line-sensitive pipelines.

**Comment by eitsupi** [comment]
I wonder this issue would be closed by #2575.

**Comment by aljazerzen** [comment]
Yes it would.
